README.md
<p align="center">
  <img src="/src/images/camera128.png" alt="Screenshot logo">
  <br>
  <b>Screenshot</b>
</p>

**Table of contents**

* <a href="#introduction"> Introduction </a>
* <a href="#features"> Features </a>
* <a href="#preview"> Preview </a>
* <a href='#install'>Install</a>
  * <a href='#install-webstores'> Chrome Webstore </a>
  * <a href='#install-fromsource'> Source install</a>
* <a href='#dependencies'>Dependencies</a>
* <a href='#thanks'>Thanks</a>
* <a href='#license'>License</a>

## <a id='introduction'>Introduction</a>

Screenshot is an extension for Chromium-derived (eg Chrome) browsers.  
The extension can take screenshots of web pages and web videos.

## <a id='features'>Features</a>

* Take a screenshot of the page on the current tab using a keyboard shortcut  
  (`Ctrl + Shift + X` on PCs and  `Command + Shift + X` on Macs).

* Take a screenshot of the video being played on the current tab using a keyboard shortcut  
  (`Ctrl + Shift + V` on PCs and  `Command + Shift + V` on Macs).

* Open screenshots in a new tab, save screenshots to disk, and copy screenshots
  to the clipboard.

* Easy to use and doesn't intrude on your privacy.


## <a id='preview'>Preview</a>

<p align="center">
  <img src="/webstore-assets/readme-screenshot.png">
</p>

## <a id='install'> Install </a>

<a id='install-webstores'>**Chrome Webstore**</a>

Screenshot is available for install via the [Chrome Webstore](https://chrome.google.com/webstore/detail/screenshot/ehmcpclingghgidajkpodncclbginiak).

<a id='install-fromsource'>**Source install**</a>

* Clone a copy of the extension:

      git clone https://github.com/rg-3/screenshot.js

* Convert the SCSS files to CSS
  (This step requires the `scss` executable to be in `$PATH`):

      cd screenshot.js
      sh scripts/build.sh

* Open `chrome://extensions` in your browser.

* Enable the `Developer mode` checkbox.

* Click the `Load unpacked extension` button and point it at 
  the `src/` directory in the cloned repository.

* Done.

## <a id='dependencies'> Dependencies </a>

**Runtime dependencies**

Dependencies used while the extension is running:

* [Spectre.css](https://picturepan2.github.io/spectre/)  
  A Lightweight, Responsive and Modern CSS Framework.

* [Popper.js](https://popper.js.org/), [Tippy.js](https://atomiks.github.io/tippyjs/)  
  Provides popovers and tooltips.

**Development dependencies**

Dependencies used to develop the extension:

  * [Saas](https://www.sass-lang.com)  
    Saas extends CSS with useful features.

## <a id='thanks'>Thanks</a>

Thanks to the following [freeicons.io](https://freeicons.io) artists:

* Thanks to [Raj Dev](https://freeicons.io/profile/714) who 
  authored the camera icon, the save icon, the copy icon and 
  the delete icon.

* Thanks to [BECRIS](https://freeicons.io/profile/3484) who 
  authored the back icon.

* Thanks to [ColourCreatype](https://freeicons.io/profile/5790) who 
  authored the settings icon.
  
 * Thanks to [Anu Rocks](https://freeicons.io/profile/730) who 
  authored the help icon.
  
## <a id='license'>License</a>

MIT license. See [./src/LICENSE.txt](./src/LICENSE.txt) for details.


CHANGELOG.md
# CHANGELOG

## next

* spectre.css: remove use of deprecated -webkit-details-marker for 
  marker, and update to v0.5.9.

* footer: do not highlight the footer images upon selection.

**v0.16.2**

* tooltips: `Delete screenshot from memory` => `Delete screenshot`.

**v0.16.1**

* Add spacing between the action bar icons

* Replace `feather.min.js` with the use of the `<img>` tag and SVGs.

**v0.16.0**

* Add settings to control whether or not tooltips are shown on the 
  settings and help icons.

* Change all tippy-controlled content to use `light-border` theme.

* Show tooltip when hovering over settings / help icons.

* Replace footer icons.

**v0.15.1**

* Update project description in `src/manifest.json`

**v0.15.0**

* End the emoji experiment.

**v0.14.8**

* Replace emoji used when help popover is open

**v0.14.7**

* Show Happy emoji by default.

**v0.14.6**

* Improve layout of `src/`.

* Update `app.maxScreenshots` before potentially long operation on 
  the setting page.

**v0.14.5**

* Fix bug(s) related to the the infinity option and the settings page.

**v0.14.4**

* Drastically improve performance of the `Delete` operation 
  by not redrawing the whole grid.

* Fix alignment of Screenshot text.

**v0.14.3**

* Represent infinity with 0 and avoid enforcing a size limit 
  on the `screenshots` array when maxScreenshots is equal to 0.

**v0.14.2**

* Reduce whitespace between footer and last two screenshots.

* Change options for number of screenshots to store in memory to:
  4, 8, 16, 32, 64, 99, Infinity

**v0.14.1**

* Fix CSS bug that caused the UI to cut off the action icons 
  for the last two screenshots. This occurred when there was 
  8 or more screenshots.

**v0.14.0**

* Redesign the navbar icons and in the process switch to a simpler
  implementation. We're not using feathericons for the navbar icons
  any longer.

**v0.13.2**

* Re-implement getFilename on Screenshot objects in graw-grid.js

**v0.13.1**

* Offer to save the most recent screenshot as 'Screenshot 4.png'
  instead of 'Screnshot 1.png' - assuming there's four screenshots.

**v0.13.0** 

* Redesign the footer to have a smaller height.

* Replace logo font KronaOne with NerkoOne.

**v0.12.1**

* Decrease infinity from 2000 to 999.

**v0.12.0**

* Add "∞" (infinity) as an option in the dropdown used to determine 
  how many screenshots to store in memory. Infinity is not infinite
  but represented by the number 2000, which should be a large enough 
  number to give the feeling of infinity. 
  
  Worth mentioning that performance could be a problem if there's not 
  enough RAM to accomodate 2000 screenshots comfortably. 

**v0.11.3**

* Align grid and settings page height when there's 4 screenshots 
  on the grid page.

**v0.11.2**

* Remove arrow from help popover by using the `arrow` prop.

* Increase height between help icon and its popover by using the `offset` prop.

* Upgrade Popper.js and Tippy.js

**v0.11.1**

* Increase the width and font size of the help popover.

**v0.11.0**

* Change logo font to KronaOne Regular.

* Change default font to Roboto Regular.

* Add new extension icon sourced from [https://freeicons.io](freeicons.io)

* Change license year in `src/LICENSE.txt`.

**v0.10.3**

* Remove underline that appears after clicking save button.

* Change default screenshot filename to `Screenshot X.png`, where X is the
  screenshot index number.

**v0.10.2**

* Improve UI and UX when there are more than 4 screenshots.

__v0.10.1__

* Fix error during read of settings.

__v0.10.0__

* Settings: persist settings between browser restarts.

* Move `LICENSE.txt` to `src/LICENSE.txt`.

* Change "Store the last X screenshots in temporary browser memory" to
  "Store the last X screenshots" on the settings page.

__v0.9.4__

* Drop `.js` suffix from the README and extension UI.

__v0.9.3__

* Add `extraHeaders` to `opt_extraInfoSpec` in order to be able to
  set the `access-control-allow-origin` header successfully. This
  appears to have been required since Chrome 79.

* Add an updated copy of [`rstl.min.js`](https://github.com/rg-3/rstl.js), and
  link to it through a `<script>` tag.

__v0.9.0__

* Halve the width of the extension along with other UI improvements.

* Delete all properties belonging to a Screenshot object after the objects
  blob is revoked using `revokeBlob()`.

* Improve device recognition settings notification.  
  See [#6](https://github.com/rg-3/screenshot.js/issues/6) for more info.

* Resort `src/html`, `src/js`, `src/css`:  

  * The contents of `src/js/lib/` have been moved to `src/js/`.

  * `browser_action.html` renamed to `grid-page.html`.

  * `browser_action.css` renamed to `grid-page.css`.

  * `settings.html` renamed to `settings-page.html`.

  * `settings.css` renamed to `settings-page.css`.


* Change the max screenshots drop down options to:  
  4, 8, 12, 16, 20, 24, 28, 32.

* Move the remaining CSS files in `src/css` to `src/scss`.  
  Keep `src/css` for the CSS files that are generated from `src/scss`.

* Numerous improvements to the README and inline documentation in the
  source code.

__v0.8.0__

* Change the default number of screenshots stored in memory from 8 to 4.
  See https://github.com/rg-3/screenshot.js/issues/3 for more info.

* A loading spinner is shown until the screenshot grid begins to get drawn by
  the `grawGrid()` function. This change avoids a user seeing an empty white
  page for 1 or 2 seconds while the grid is waiting to be drawn, which occurs
  after HTML blobs are inserted by `includeHTML()`. Sometimes there's no wait,
  it seems to depend on the number of screenshots waiting to be processed.

* The settings and help icons are fixed in place, they follow scroll when
  there is more than 8 screenshots to browse, and they've been positioned to
  the center.

* Add a settings page that allows the user configure the extension.  
  (The two options that can be configured are described below).


src/js/settings-page.js
import includeHTML from "./browser-action/include-html.js";
import tippyFooter     from './browser-action/tippy-footer.js';
import drawCommandHelp from "./browser-action/draw-command-help.js";

const setDefaultOption = (select, currentValue) => {
  for(let i = 0; i < select.options.length; i++) {
    const option = select.options[i];
    if(option.value === String(currentValue)) {
      option.selected = true;
    }
  }
};

chrome.runtime.getBackgroundPage((page) => {
  const app = page.app;

  app.getKeyboardCommands().then((commands) => {
    includeHTML().then((nodes) => {
      commands.forEach(drawCommandHelp);
      tippyFooter(app);
    });
  });

  /*
     Control the size of a video screenshot.
     Possible values: 'visible', 'natural'.
  */
  const vSizeSelect = document.getElementById('video-size');
  setDefaultOption(vSizeSelect, app.settings.getItem('videoSize'));
  vSizeSelect.addEventListener('change', (event) => {
    const option = event.target;
    app.settings.setItem('videoSize', option.value);
  });

  /*
    Control the number of screenshots stored in memory.
    Possible values: 4, 8, 16, 32, 64, 99, Infinity
  */
  const maxSelect = document.getElementById('max-screenshots');
  setDefaultOption(maxSelect, app.settings.getItem('maxScreenshots'));
  maxSelect.addEventListener('change', (event) => {
    const option   = event.target;
    const newMax   = Number(option.value);
    const truncate = app.maxScreenshots === 0 || (newMax > 0 && newMax < app.maxScreenshots);
    app.settings.setItem('maxScreenshots', newMax);
    if(truncate) {
      for(let i = newMax; i < app.screenshots.length; i++) {
        const screenshot = app.screenshots[i];
        chrome.runtime.sendMessage({action: 'remove-screenshot', removedId: screenshot.id});
      }
    }
  });

  /*
    Control whether or not tooltips are shown on the settings and help icons.
    Possible values: 1 (show), 0 (hide)
  */
  const tooltipSelect = document.getElementById('show-tooltips');
  setDefaultOption(tooltipSelect, app.settings.getItem('showTooltips'));
  tooltipSelect.addEventListener('change', (event) => {
    const option = event.target;
    app.settings.setItem('showTooltips', Number(option.value));
  });

    /*
      Control whether or not a sound is played when a screenshot is taken.
      Possible values: 1 (show), 0 (hide)
    */
    const playSoundSelect = document.getElementById('play-sound');
    setDefaultOption(playSoundSelect, app.settings.getItem('playSound'));
    playSoundSelect.addEventListener('change', (event) => {
      const option = event.target;
      app.settings.setItem('playSound', Number(option.value));
    });
});
